File: C/Exercice4_4_pere.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main()
{
    int tube[2];
    __pid_t fils;
    char tube_str[10],
         message[] = "Bonjour depuis le pere !";

    if(pipe(tube) != 0)
    {
        puts("Erreur lors de pipe\n");
        exit(0);
    }
    else
    {
        fils = fork();
        if(fils == -1)
        {
            puts("Erreur lors de fork\n");
            exit(0);
        }
        if(fils == 0)
        {
            close(tube[1]);
            sprintf(tube_str, "%d\n", tube[0]);
            setenv("TUBE_FD", tube_str, 1);
            char *args[] = {"./fils", tube_str, NULL};
            execv("./fils", args);
            exit(1);
        }
        else
        {
            close(tube[0]);

            write(tube[1], message, sizeof(message));
            close(tube[1]);
        }
    }

    return 0;
}

File: Python/Exercice4_4_pere.py
#coding: utf-8
import os
import sys

r, w = os.pipe()

os.set_inheritable(r, True)

fils = os.fork()

if fils == -1:
    print("Erreur lors de fork\n")
    sys.exit(1)

if fils == 0:
    os.close(w)
    os.environ["TUBE_FD"] = str(r)

    os.execv(sys.executable, [sys.executable, "Exercice4_4_fils.py"])

    sys.exit(1)
else:
    os.close(r)
    message = "Bonjour depuis le pere !"
    os.write(w, message.encode())
    os.close(w)